### PR TITLE
Add newton.usd.runtime for stage-driven simulation

### LINKED ISSUE #3114: [REQ] Newton USD Runtime for Data-Driven Simulation
## Newton USD Runtime for Data-Driven Simulation

  Build the Newton USD runtime entry points that allow a user or application to load a USD-authored simulation, construct the required Newton assets/solvers/models/couplers, advance the simulation, and optionally write updated state back to the runtime stage (tbd). The goal is that the model is the spec: a user can provide a stage, initialize Newton from it, step the simulation, and observe results without manually wiring bodies, shapes, solvers, contacts, or data updates.

  ### Scope

  - Define and implement the core runtime API surface:
    - `newton.usd.runtime.load_usd()` — load stage content and derive Newton assets, solvers, models, and couplers.
    - `newton.usd.runtime.step()` — advance the stage-driven simulation and all associated solvers.
    - `newton.usd.runtime.fetchresults()` — optionally update the runtime stage (tbd) with the latest simulated state.
  - Clarify how this runtime relates to existing simUSD code and reuse/merge existing functionality where appropriate.
  - Support data-driven configuration from USD-authored physics content, including bodies, shapes, transforms, solver setup, and future coupling descriptions.
  - Identify the required data-access interfaces for reading stage state and writing simulation results.
  - Keep the interface suitable for downstream application integration and agent-driven workflows.

  ### Deliverables

  - A minimal working Newton USD runtime path that can load a representative stage, initialize Newton state, step simulation, and publish updated transforms/state.
  - API documentation for `load_usd()`, `step()`, and `fetchresults()`, including expected inputs, lifecycle, and ownership assumptions.
  - One runnable example showing stage-driven Newton simulation without manual Newton scene construction.
  - A short architecture note covering relation to simUSD, runtime stage (tbd), data-access assumptions, and open gaps.
  - A follow-up issue list for missing schema support, data-plane gaps, coupling support, and sensor/annotation integration needs.

  ### Out of Scope

  - New solver development.
  - Full productization of every USD physics schema.
  - Renderer or sensor implementation work beyond exposing the Newton-side data needed by those systems.
  - Finalizing the long-term runtime stage architecture if it is still undecided.

###  Historical context / prior work

  There was substantial earlier SimUSD work in Newton, although the complete runtime was never merged.

  - #636 introduced internal `sim_usd.py` / `sim_usd_gtc.py` prototypes. Their `Simulator` objects loaded USD, selected/configured a solver, owned Newton states and contacts, stepped the simulation, and wrote results back to USD. The PR remained a draft and was closed for
  inactivity without merging.
  - #201 separately merged the USD write-back/rendering portion originally used by SimUSD. This evolved into today’s `ViewerUSD`.
  - #626 merged the public `ModelBuilder.add_usd()` entry point and improved USD visual-shape import.
  - #710 merged the schema-resolver and custom-attribute infrastructure, including scene timing, solver iteration metadata, and solver-specific attributes.

  Current Newton therefore contains much of the substrate, but not the complete runtime:

  - `newton.usd.runtime` does not currently exist.
  - `ModelBuilder.add_usd()` constructs model content and returns import metadata, but it does not select a solver or own states, contacts, control, or stepping.
  - `newton.usd` currently exposes USD utilities and schema resolvers.
  - `ViewerUSD` provides time-sampled USD output.

  The proposed runtime should treat #636 as prior art rather than restore it wholesale. The useful central idea is that each loaded simulation owns its model, selected solver(s), state buffers, control, contacts, timestep, and stepping policy. The earlier prototype also
  mixed in CLI, rendering, application-specific BDX helpers, duplicated schema mappings, and now-obsolete Newton APIs, so those concerns should remain separate.

  Relevant links:

  - Prior SimUSD prototype: #636
  - USD write-back extraction: #201
  - `ModelBuilder.add_usd()`: #626
  - Schema resolvers and custom attributes: #710

### PR DESCRIPTION
**Pull request by adenzler-nvidia**
## Description

Add `newton.usd.runtime`: stage-driven simulation where the USD stage is the spec. `load_usd(source)` derives the model, solver, states, control, collision pipeline, contacts, and timestep from a stage; `sim.step()` advances one stage-authored timestep. No Python scene wiring is required.

Closes #3114

```python
sim = newton.usd.runtime.load_usd("scene.usda")
sim.step()
viewer.log_state(sim.state)   # viewers consume state directly; no publish API
```

Two deliberate divergences from the #3114 sketch: `step` is a method on `Simulation` rather than a module-level function (consistent with `solver.step()` / `pipeline.collide()`; `load_usd()` stays the only module-level entry point), and there is no `fetchresults()` — viewers and applications consume `sim.state` directly, with stage write-back deferred to the stage-writer follow-up.

The solver is selected by applying a `NewtonSolver<Name>API` schema to the stage's single `PhysicsScene` prim; solver parameters live in `newton:<solver>:*` attributes on the same prim:

```usda
def PhysicsScene "physicsScene" (
    prepend apiSchemas = ["NewtonSolverXpbdAPI"]
)
{
    float newton:timeStepsPerSecond = 480
    int newton:xpbd:iterations = 4
}
```

Key semantics:

- Selection by applied API schema (XPBD, MuJoCo, Featherstone, SemiImplicit, VBD). The schemas are intentionally unregistered for now and read via `has_applied_api_schema()` (the existing draft-schema pattern); stages authored today validate unchanged once codeless schemas ship. Until then `usdchecker` warns about the unregistered API — accepted intermediate cost, first follow-up.
- Parameters are authored-only with solver-constructor defaults; unrecognized attributes are inert, like any custom data — no warnings, no errors. MuJoCo options keep their existing single channel (`mjc:option:*` custom attributes); the registry exposes only `njmax`/`nconmax`, which have no other spelling.
- Timestep resolves through the schema-resolver stack, so PhysX (`physxScene:timeStepsPerSecond`) and MuJoCo (`mjc:option:timestep`) spellings work without Newton-specific authoring.
- `Simulation.step()` is in-place (`state_in is state_out`, the established pattern), clears forces after the solve (user forces written between steps act for exactly one step), and collides on a `newton:collisionInterval` cadence. Graph capture (CUDA and CPU) is on by default with a `use_graph` override; capture never advances the sim (test-enforced bitwise).
- All validation happens at load; the source stage is never written; everything on `Simulation` is a public field, so external control and gradient workflows (`requires_grad=True`) opt in by using the components directly.
- A parity test locks the no-hidden-physics guarantee: the same stage wired manually must produce the runtime's exact trajectory.

Also included:

- `newton/usd.py` becomes the `newton/usd/` package (pure re-export move, public surface unchanged).
- Fix: `add_usd()` no longer prints spurious "missing customData" warnings for plain `newton:*` attributes; attributes with no customData are simply not custom-attribute declarations. Incomplete declarations (one of the two keys) still warn.
- CLI runner `python -m newton.usd.runtime <stage> --viewer gl|usd|null` plus the `usd_runtime` gallery example with a checked-in reference stage. Note for reviewers: #3114 lists the CLI among concerns that "should remain separate" from the runtime — it is structurally separate (an isolated `_main`; the runtime objects know nothing about it) and severable in one commit if we'd rather not ship an entry point yet.
- `docs/generate_api.py` learns to page `newton.usd` submodules (mirrors the existing `newton.solvers` handling).

Relation to prior art: this follows the #636 simUSD prototype's central idea (a loaded simulation owns its model/solver/states/timestep) while replacing its hardcoded attribute tables with the #710 schema resolvers, dropping the frame/substep time model, and keeping CLI/rendering out of the runtime objects. Write-back to source prim paths (the #636 `UpdateUsd` capability) is deliberately deferred.

Follow-ups (issues to be filed): codeless `NewtonSolver*` schemas; source-path-preserving stage writer; runtime grad mode (tape + trajectory); stage re-sync via `notify_model_changed`; time-sampled control and kinematic bodies; `num_worlds` replication; couplers (N scenes → N solvers, `physics:simulationOwner`); public solver registration; viewer player (drag-and-drop).

## Checklist

- [x] New or existing tests cover these changes
- [x] The documentation is up to date with these changes
- [x] `CHANGELOG.md` has been updated (if user-facing change)

## Test plan

```
uv run --extra dev -m newton.tests -k test_usd_runtime.TestUsdRuntime
uv run --extra dev -m newton.tests -k "test_examples.TestUsdExamples"
uv run --extra dev python -m newton.usd.runtime newton/examples/assets/usd_runtime_boxes.usda --viewer null --num-steps 100
```

Runtime suite covers selection errors, parameter mapping and defaults, source-stage immutability, capture-does-not-advance, graph/direct trajectory parity, manual-wiring parity, per-solver load-and-step for all five registered solvers, and the CLI. Full suite verified at the branch head (only pre-existing hydroelastic failures, reproduced bit-identically at the base commit).

## Summary by CodeRabbit

* **New Features**
  * Added a stage-driven USD simulation runtime with `load_usd` and `step`, including optional graph-capture support and a new “USD Examples” entry plus a runnable USD runtime example and asset.
* **Bug Fixes**
  * Improved handling of USD `customData` for `newton:` attributes to eliminate spurious warnings and better skip incomplete/inert attributes.
* **Documentation**
  * Added/expanded API docs for the USD runtime, updated API generation to include USD submodules, and updated module exports.
* **Tests**
  * Added unit tests for USD runtime loading/stepping, graph-capture behavior, and a new example-based test.

### CODE REVIEW

File: newton/examples/usd/example_usd_runtime.py [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+# SPDX-License-Identifier: Apache-2.0
+
+###########################################################################
+# Example USD Runtime
+#
+# Stage-driven simulation: the USD file authors the physics scene, the
+# solver choice, and the solver configuration. No manual scene wiring.
+#
+# Command: python -m newton.examples usd_runtime
+###########################################################################
```
[the remark this answers is not in the document]
Comment (adenzler-nvidia) [reply newton/examples/usd/example_usd_runtime.py:11]: This banner is the repo-wide example convention — every file under newton/examples carries the same header format (description + run command), so keeping it for consistency.

File: newton/tests/test_usd_runtime.py [hunk 2]
```diff
@@ -0,0 +1,352 @@
...
+
+    def test_all_registered_solvers_load_and_step(self):
+        import newton.usd.runtime as runtime  # noqa: PLC0415
+        from newton._src.usd.runtime import _SOLVER_REGISTRY  # noqa: PLC0415
+
+        for schema, entry in _SOLVER_REGISTRY.items():
+            with self.subTest(solver=schema):
+                if entry.cls is newton.solvers.SolverMuJoCo and not MUJOCO_AVAILABLE:
+                    self.skipTest("Requires mujoco-warp")
+                sim = runtime.load_usd(_make_stage(solver_api=schema), use_graph=False)
```
[the remark this answers is not in the document]
Comment (adenzler-nvidia) [reply newton/tests/test_usd_runtime.py:291]: The skipTest call sits inside the `with self.subTest(...)` block, where unittest catches SkipTest per subtest and continues the loop — the remaining registry entries still run. Verified: with mujoco-warp absent, featherstone/semiImplicit/vbd subtests execute.

**Comment by jcarius-nv** [comment]
Related: https://github.com/newton-physics/newton/pull/3629

**Comment by jcarius-nv** [comment]
General thought here: We usually face the requirement that physics stepping should remain independent and asynchronous from sensor evaluation and other upstream or downstream operations (e.g., rendering).

In my opinion the single most important piece to enable this without generating significant complexity inside of newton.usd.runtime is to keep the state object as explicitly user-controlled. That way users can decide how many simulation steps they want to run in parallel, can manage their own circular state buffer if required, evaluate sensors at a later time, etc.
So contrary to the PR description (at the time of writing this), I think it might simplify things a lot if we'd have it like

```python
sim = newton.usd.runtime.load_usd("scene.usda")
state0 = sim.state()
state1 = sim.state()
sim.step(state0, state1)

viewer.log_state(state1)   # viewers consume state directly; no publish API
more_work_consuming_state(state1)

sim.step(state1, state2) # state1 read-only, no danger of mutating it while above viewer/rendering/sensor/... operations are still inflight
```

Further, we need to keep in mind that the model may be constructed by other means that a USD file, for example from an [ovstage](https://nvidia-omniverse.github.io/ovstage/) via [ovnewton](https://github.com/NVIDIA-Omniverse/ovnewton/tree/main/), so there must be a way to obtain the runtime `sim` object also in that case.
